src/corelibs/U2View/src/ov_sequence/view_rendering/DetViewSingleLineRenderer.cpp

```diff
@@ -419,112 +411,96 @@ void DetViewSingleLineRenderer::drawTranslations(QPainter &p, const U2Region &vi
-                                                       const char *seqBlock,
-                                                       const QList<SharedAnnotationData> &annotationsInRange) {
-    bool isTranslateAnnotationOrSelection = (ctx->getTranslationState() == SequenceObjectContext::TS_AnnotationsOrSelection);
-    QVector<U2Region> regions;
-    QList<QVector<U2Region>> sortedRegions = QList<QVector<U2Region>>() << QVector<U2Region>() << QVector<U2Region>() << QVector<U2Region>();
-    if (isTranslateAnnotationOrSelection) {
-        regions = ctx->getSequenceSelection()->getSelectedRegions();
-        QList<int> mods;
-        for (int i = 0; i < 3; i++) {
-            trMetrics.visibleRows[i] = false;
+                                                       const char *visibleSequence,
+                                                       const QList<SharedAnnotationData> &annotationsInRange,
+                                                       const TranslationMetrics &translationMetrics) {
+    DNATranslation *aminoTT = detView->getAminoTT();
+    CHECK(aminoTT != nullptr, );
+
+    int lineIndex = firstDirectTransLine;
+    for (int frameIndex = 0; frameIndex < 3; frameIndex++) {
+        if (!translationMetrics.visibleFrames[frameIndex]) {
+            continue;
         }
-        foreach (const U2Region &reg, regions) {
-            const int mod = reg.startPos % 3;
-            trMetrics.visibleRows[mod] = true;
-
-            const int intersectedLength = visibleRange.intersect(reg).length;
-
-            if (intersectedLength > (reg.length % 3) + 1) {
-                sortedRegions[mod] << reg;
-                if (!mods.contains(mod)) {
-                    mods << mod;
+        int visibleRangeFrameIndex = visibleRange.startPos % 3;
+        qint64 translationStartPos = visibleRange.startPos + frameIndex - visibleRangeFrameIndex - 3;
+        while (translationStartPos < qMax((qint64)0, visibleRange.startPos - 1)) {
+            translationStartPos += 3;
+        }
+        qint64 translationIndent = translationStartPos - visibleRange.startPos;
+        const char *sequenceToTranslate = visibleSequence + translationIndent;
+        qint64 sequenceToTranslateLength = qMin(visibleRange.endPos() + 1, ctx->getSequenceLength()) - translationStartPos;
+        QByteArray aminoSequence = translate(aminoTT, sequenceToTranslate, sequenceToTranslateLength);
+
+        QVector<bool> aminoVisibilityFlags(aminoSequence.size(), true);    // Visible amino acid positions. All positions are visible by default.
+        if (ctx->getTranslationState() == SequenceObjectContext::TS_AnnotationsOrSelection) {
+            aminoVisibilityFlags.fill(false);    // Mark as visible only acids within selected regions.
+            U2Region translatableRange(visibleRange.startPos + translationIndent, aminoSequence.length() * 3);
+            for (auto selectedRegion : ctx->getSequenceSelection()->getSelectedRegions()) {
```

Comment:
for (const auto& selectedRegion : ctx->getSequenceSelection()->getSelectedRegions()) {

Reply:
Fixed